### DISCUSSION TITLE
Cross-Platform Input Support and Integration with Tnua

### DISCUSSION THREAD

**Discussion by heydocode** [idanarye/bevy-tnua#65] [Q&A]
### **Cross-Platform Input Support and Integration with Tnua**

Hello,

I am currently developing a **cross-platform game** and looking to implement an **input manager** that can handle **sensor**, **gamepad**, and **keyboard** inputs seamlessly across platforms (desktop, mobile, etc.).

I have a couple of questions regarding the Tnua crate's capabilities and flexibility with respect to input handling:

1. **Does Tnua support sensor, gamepad, and keyboard inputs natively**, or will I need to handle these input types separately and feed them into Tnua's movement functions?

2. **Is it possible to integrate Tnua with a custom input system**? For example, I am considering using the [Virtual Joystick plugin](https://github.com/SergioRibera/virtual_joystick) for mobile input. Can Tnua receive input from this (or similar) plugins to trigger movement actions like **forward walk**, **jump**, etc.? Does Tnua expose functions or events for actions such as walking, jumping, or dashing that could be fed these inputs?

### **Context:**
I am building a game that needs to support input from **multiple devices** (touch on mobile, gamepads, keyboard on desktop, etc.), and I am wondering if there is already an input management solution compatible with Tnua or if I should create one myself. Essentially, I'd like to know how easily I can hook up inputs from different devices and sensors into Tnua's movement system.

Any guidance or information would be appreciated!

Thank you for your time.

**Comment by idanarye** [comment]
Tnua does not handle player input at all - even if you only support a keyboard, you are still expected to read the input yourself and call the methods of [`TnuaController`](https://docs.rs/bevy-tnua/0.19.0/bevy_tnua/controller/struct.TnuaController.html) in order to feed it into the character controller. So it doesn't matter where the input comes from - keyboard, gamepad, joytstick, touchscreen, or even some NPC AI controller.

Refer to the example to see how it's done: https://github.com/idanarye/bevy-tnua/blob/3562f35ef7c186c84d19b4855f8d3452e397839f/examples/example.rs#L130

**Comment by idanarye** [comment]
> and I am wondering if there is already an input management solution compatible with Tnua

Not really "compatible with Tnua" because Tnua does not need compatibility, but I believe the community standard is [Leafwing Input Manager](https://github.com/Leafwing-Studios/leafwing-input-manager). This is what I use, and it works nice with Tnua. It supports keyboard, gamepad, and mouse - but I don't think it supports touchscreen (and I don't even know what you mean by "sensor"). I don't know if you can extend it to add support for these yourself.

**Comment by heydocode** [comment]
By sensor I meant touchscreen input support. Thanks for help!
